src/rdkafka_ssl.c

```diff
@@ -1174,6 +1175,74 @@ static int rd_kafka_ssl_set_certs (rd_kafka_t *rk, SSL_CTX *ctx,
                 check_pkey = rd_true;
         }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        /*
+         * If applicable, use OpenSSL engine to fetch SSL certificate.
+         */
+        if (rk->rk_conf.ssl.engine) {
+                STACK_OF(X509_NAME) *cert_names = sk_X509_NAME_new_null();
+                STACK_OF(X509_OBJECT) *roots = 
+                    X509_STORE_get0_objects(SSL_CTX_get_cert_store(ctx));
+                X509 *x509 = NULL;
+                EVP_PKEY *pkey = NULL;
+                int i = 0;
+                for (i = 0; i < sk_X509_OBJECT_num(roots); i++) {
+                        x509 = 
+                            X509_OBJECT_get0_X509(sk_X509_OBJECT_value(roots, i));
+                        if (x509)
+                                sk_X509_NAME_push(cert_names, 
+                                                  X509_get_subject_name(x509));
+                }
+
+                x509 = NULL;
+                r = ENGINE_load_ssl_client_cert(rk->rk_conf.ssl.engine, 
+                                                NULL,
+                                                cert_names,
+                                                &x509,
+                                                &pkey,
+                                                NULL, NULL, 
+                                                rk->rk_conf.ssl.engine_callback_data);
+
+                sk_X509_NAME_free(cert_names);
+                if (r == -1 || !x509 || !pkey) {
+                        X509_free(x509);
```

Comment:
Are these calls handling NULL pointers? If not, use RD_IF_FREE()

Reply:
both EVP_PKEY_free X509_free handle nulls